## Data Science Module 1 Final Project: <br>Predicting the Sale Price of Houses in King County

Student Names: Natasha Kacoroski and David Kaspar 
Student Pace: Full time 
Scheduled Project Review Date/Time: Unknown 
Instructor Names: Miles Erickson and Greg Damico 
Blog Post URL: Unknown

Before doing anything with the data, we need to import the python libraries that we will be using to manipulate the data (explain what libraries and why) and import the data into a pandas dataframe.

In [1]:
import pandas as pd
import numpy as np
import statsmodels.api as sm
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import sklearn as sk
from scipy.stats import zscore
%matplotlib inline

kch_df = pd.read_csv('kc_house_data.csv')
df = kch_df.copy()

## Step 1: Clean King County Housing Data

### Initial data observations

The first step of data cleaning is to take an initial look at the dataset and see what it looks like, using transpose allows us to see all the columns easily.

In [2]:
df.head().T

0           1           2           3           4
id             7129300520  6414100192  5631500400  2487200875  1954400510
date           10/13/2014   12/9/2014   2/25/2015   12/9/2014   2/18/2015
price              221900      538000      180000      604000      510000
bedrooms                3           3           2           4           3
bathrooms               1        2.25           1           3           2
sqft_living          1180        2570         770        1960        1680
sqft_lot             5650        7242       10000        5000        8080
floors                  1           2           1           1           1
waterfront            NaN           0           0           0           0
view                    0           0           0           0           0
condition               3           3           3           5           3
grade                   7           7           6           7           8
sqft_above           1180        2170         770        1050        1680
sqft_basement         0.0       400.0         0.0       910.0         0.0
yr_built             1955        1951        1933        1965        1987
yr_renovated            0        1991         NaN           0           0
zipcode             98178       98125       98028       98136       98074
lat               47.5112      47.721     47.7379     47.5208     47.6168
long             -122.257    -122.319    -122.233    -122.393    -122.045
sqft_living15        1340        1690        2720        1360        1800
sqft_lot15           5650        7639        8062        5000        7503

Looking at the head of the dataframe, we can see that the columns are the appropriate data labels not the first entry. To confirm and check the column information we imported the .md file on column information.

In [3]:
from IPython.display import display, Markdown

with open('column_names.md', 'r') as fh:
    content = fh.read()

display(Markdown(content))

# Column Names and descriptions for Kings County Data Set
* **id** - unique identified for a house
* **dateDate** - house was sold
* **pricePrice** -  is prediction target
* **bedroomsNumber** -  of Bedrooms/House
* **bathroomsNumber** -  of bathrooms/bedrooms
* **sqft_livingsquare** -  footage of the home
* **sqft_lotsquare** -  footage of the lot
* **floorsTotal** -  floors (levels) in house
* **waterfront** - House which has a view to a waterfront
* **view** - Has been viewed
* **condition** - How good the condition is ( Overall )
* **grade** - overall grade given to the housing unit, based on King County grading system
* **sqft_above** - square footage of house apart from basement
* **sqft_basement** - square footage of the basement
* **yr_built** - Built Year
* **yr_renovated** - Year when house was renovated
* **zipcode** - zip
* **lat** - Latitude coordinate
* **long** - Longitude coordinate
* **sqft_living15** - The square footage of interior housing living space for the nearest 15 neighbors
* **sqft_lot15** - The square footage of the land lots of the nearest 15 neighbors


Looking at the column information, most columns descriptions confirm our initial understanding of the column name. We learn that view means whether the house has been viewed (assumed to be potential buyers), and sqft_living15 and sqft_lot15 refers to the square footage of interior housing living space and land lots for nearest 15 neighbors, respectively. Also of interest is the grade given to housing unit based on King County grading system. We looked up the King County grading system here: <br> https://info.kingcounty.gov/assessor/esales/Glossary.aspx?type=r <br>
to understand the grading system.

BUILDING GRADE <br>
Represents the construction quality of improvements. Grades run from grade 1 to 13. Generally defined as:

1-3 = Falls short of minimum building standards. Normally cabin or inferior structure.

4 = Generally older, low quality construction. Does not meet code.

5 = Low construction costs and workmanship. Small, simple design.

6 = Lowest grade currently meeting building code. Low quality materials and simple designs.

7 = Average grade of construction and design. Commonly seen in plats and older sub-divisions.

8 = Just above average in construction and design. Usually better materials in both the exterior and interior finish     work.

9 = Better architectural design with extra interior and exterior design and quality.

10 = Homes of this quality generally have high quality features. Finish work is better and more design quality is seen in the floor plans. Generally have a larger square footage.

11 = Custom design and higher quality finish work with added amenities of solid woods, bathroom fixtures and more luxurious options.

12 = Custom design and excellent builders. All materials are of the highest quality and all conveniences are present.

13  = Generally custom designed and built. Mansion level. Large amount of highest quality cabinet work, wood trim, marble, entry ways etc. 

After getting a better understanding of each column, we moved to looking at values within each column. We also had the initial thought of moving the id column to be the primary key if there are no duplicates. 

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
id               21597 non-null int64
date             21597 non-null object
price            21597 non-null float64
bedrooms         21597 non-null int64
bathrooms        21597 non-null float64
sqft_living      21597 non-null int64
sqft_lot         21597 non-null int64
floors           21597 non-null float64
waterfront       19221 non-null float64
view             21534 non-null float64
condition        21597 non-null int64
grade            21597 non-null int64
sqft_above       21597 non-null int64
sqft_basement    21597 non-null object
yr_built         21597 non-null int64
yr_renovated     17755 non-null float64
zipcode          21597 non-null int64
lat              21597 non-null float64
long             21597 non-null float64
sqft_living15    21597 non-null int64
sqft_lot15       21597 non-null int64
dtypes: float64(8), int64(11), object(2)
memory usage: 3.5+ MB


The shape of our data is 21,587 entries and 21 columns, with data types of float64(8), int64(11), and object(2). There are null values in waterfront, view, and yr_renovated. Columns to consider changing data types are: id from int64 to a object (value is independent of magnitude), date to a timestamp (makes it easier to manipulate), sqft_basement from object to int64 (value is dependent on magnitude), yr_built is int64 while yr_renovated is float64 is odd - needs further investigation, and zipcode from int64 to object (value is independent of magnitude). It would also be useful to know which columns contain continuous or discrete data because that will influence how we deal with null values and will be a factor in determining one-hot encoding.

### Addressing the null values and if values are continuous or discrete

In [5]:
def null_percentage_and_data_continuity(df):
    """Returns percentage of null values in data column if nulls exist 
    and whether data is likely continuous or discrete."""
    display_df = df.head(1).applymap(type)
    percent_nan = {}
    disc_or_cont = {}
    for i, col in enumerate(df):
        col_percent_nan = round(100*df[col].isna().sum() / len(df[col]), 2)
        percent_nan[col] = [col_percent_nan]
        if df[col].nunique() < 100:
            disc_or_cont[col] = ['disc']
        else:
            disc_or_cont[col] = ['cont']
    percent_nan_df = pd.DataFrame.from_dict(percent_nan, orient='columns')
    disc_or_cont_df = pd.DataFrame.from_dict(disc_or_cont, orient='columns')
    final = pd.concat([display_df, percent_nan_df, disc_or_cont_df]).T
    final.columns = ['dtype', 'percent null', 'disc or cont']
    return final

null_percentage_and_data_continuity(df)

dtype percent null disc or cont
id               <class 'int'>            0         cont
date             <class 'str'>            0         cont
price          <class 'float'>            0         cont
bedrooms         <class 'int'>            0         disc
bathrooms      <class 'float'>            0         disc
sqft_living      <class 'int'>            0         cont
sqft_lot         <class 'int'>            0         cont
floors         <class 'float'>            0         disc
waterfront     <class 'float'>           11         disc
view           <class 'float'>         0.29         disc
condition        <class 'int'>            0         disc
grade            <class 'int'>            0         disc
sqft_above       <class 'int'>            0         cont
sqft_basement    <class 'str'>            0         cont
yr_built         <class 'int'>            0         cont
yr_renovated   <class 'float'>        17.79         disc
zipcode          <class 'int'>            0         disc
lat            <class 'float'>            0         cont
long           <class 'float'>            0         cont
sqft_living15    <class 'int'>            0         cont
sqft_lot15       <class 'int'>            0         cont

Starting with waterfront, 11% of the values are null and the data appears to be discrete. Let's check the unique values.

In [6]:
print(df.waterfront.unique())
print(df.waterfront.value_counts())
for count in list(df.waterfront.value_counts()):
    print(count / sum(list(df.waterfront.value_counts()))*100)

[nan  0.  1.]
0.0    19075
1.0      146
Name: waterfront, dtype: int64
99.24041413037823
0.7595858696217679


Waterfront is binary, where 0 is no waterfront view and 1 means that there is a waterfront view. Null values are if waterfront view is unknown. Since the value count of 0 is significantly more common (99.2%), we decided to replace all null values with 0.

In [7]:
df.waterfront = df.waterfront.fillna(0.0)
df.waterfront.isna().any()

False

Moving onto view. The description for view is "has been viewed." Although we cannot confirm, we picture this as the number of people who have viewed the house at the time that the data for this house sale was collected. 0.29% of the view values are null and it appears to be discrete. Below we check the unique values.

In [8]:
print(df.view.unique())
print(df.view.value_counts())
print(df.view.median())
for count in list(df.view.value_counts()):
    print(count / sum(list(df.view.value_counts()))*100)

[ 0. nan  3.  4.  2.  1.]
0.0    19422
2.0      957
3.0      508
1.0      330
4.0      317
Name: view, dtype: int64
0.0
90.19225410977988
4.444134856505991
2.3590600910188537
1.5324602953468933
1.4720906473483792


Since the value count of 0 is significantly more common (90.2%, also the median), we decided to replace all null values with 0.

In [9]:
df.view = df.view.fillna(0.0)
df.view.isna().any()

False

For yr_renovated, the values are the year where any renovations took place. They appear to be discrete and 17.8% of the values are null. Below are the unique values.

In [10]:
print(df.yr_renovated.value_counts().head())
renovated_num = df.loc[df['yr_renovated'] > 0, :].shape[0]
renovated_num/len(df)*100

0.0       17011
2014.0       73
2003.0       31
2013.0       31
2007.0       30
Name: yr_renovated, dtype: int64


3.4449229059591606

Looking at the unique values we see that null and 0 are values for year renovated. We were already aware of the nulls, and decide that 0 could either mean that house was never renovated or renovations were not tracked. We decided to drop year renovated because of the uncertainty and because actual house renovations are only 3.4% of the year renovated values. 

In [11]:
df = df.drop('yr_renovated', axis=1)
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

## Change data types

For the ID column we considered changing the values to an object and making that the primary key. From looking at unique values for the column (see below), we saw that there were some duplicates because the ID seems to refer to house sale instance, and some houses have been sold more than once. Since the ID column is not valuable to us as a primary key, we decided to drop the column. We also considered dropping duplicate values, but since we are using data about all house sale instances, we decided to keep it in.

In [12]:
df.id.duplicated().any()

True

In [13]:
df = df.drop('id', axis=1)
df.columns

Index(['date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'zipcode', 'lat', 'long', 'sqft_living15',
       'sqft_lot15'],
      dtype='object')

To change the sqft_basement from an object to an integer, we tried using pd.to_numeric.

In [15]:
df.sqft_basement = pd.to_numeric(df['sqft_basement'])

ValueError: Unable to parse string "?" at position 6

And learn that at least one entry has a "?" instead of a number. Looking at the unique values we can see that a "?" is the only non-numeric value. For larger datasets with more unique values we would need to write a function to check for this.

In [17]:
df.sqft_basement.unique()

array(['0.0', '400.0', '910.0', '1530.0', '?', '730.0', '1700.0', '300.0',
       '970.0', '760.0', '720.0', '700.0', '820.0', '780.0', '790.0',
       '330.0', '1620.0', '360.0', '588.0', '1510.0', '410.0', '990.0',
       '600.0', '560.0', '550.0', '1000.0', '1600.0', '500.0', '1040.0',
       '880.0', '1010.0', '240.0', '265.0', '290.0', '800.0', '540.0',
       '710.0', '840.0', '380.0', '770.0', '480.0', '570.0', '1490.0',
       '620.0', '1250.0', '1270.0', '120.0', '650.0', '180.0', '1130.0',
       '450.0', '1640.0', '1460.0', '1020.0', '1030.0', '750.0', '640.0',
       '1070.0', '490.0', '1310.0', '630.0', '2000.0', '390.0', '430.0',
       '850.0', '210.0', '1430.0', '1950.0', '440.0', '220.0', '1160.0',
       '860.0', '580.0', '2060.0', '1820.0', '1180.0', '200.0', '1150.0',
       '1200.0', '680.0', '530.0', '1450.0', '1170.0', '1080.0', '960.0',
       '280.0', '870.0', '1100.0', '460.0', '1400.0', '660.0', '1220.0',
       '900.0', '420.0', '1580.0', '1380.0', '475.0', 

In [22]:
print(df.sqft_basement.value_counts().head())
questionmark_num = df.loc[df['sqft_basement'] == '?', :].shape[0]
questionmark_num/len(df)*100

0.0      12826
?          454
600.0      217
500.0      209
700.0      208
Name: sqft_basement, dtype: int64


2.1021438162707784

Although the questionmark has the second highest value count, that is only 2.1% of the sqft_basement values, so we decided to convert all the questionmark values to the median value because that is the measure of centrality that is least likely to skew our data.  

In [ ]:
basement_med = pd.to_numeric(df['sqft_basement'].loc[df['sqft_basement'] != '?']).median()
print(basement_med)
df.sqft_basement = df.sqft_basement.replace('?', basement_med)

In [29]:
df.sqft_basement = pd.to_numeric(df['sqft_basement'])
df.sqft_basement.dtype

dtype('float64')

Moving on to zipcode, all the unique values look like reasonable zipcodes, so we convert them to strings.

In [32]:
df.zipcode.unique()

array(['98178', '98125', '98028', '98136', '98074', '98053', '98003',
       '98198', '98146', '98038', '98007', '98115', '98107', '98126',
       '98019', '98103', '98002', '98133', '98040', '98092', '98030',
       '98119', '98112', '98052', '98027', '98117', '98058', '98001',
       '98056', '98166', '98023', '98070', '98148', '98105', '98042',
       '98008', '98059', '98122', '98144', '98004', '98005', '98034',
       '98075', '98116', '98010', '98118', '98199', '98032', '98045',
       '98102', '98077', '98108', '98168', '98177', '98065', '98029',
       '98006', '98109', '98022', '98033', '98155', '98024', '98011',
       '98031', '98106', '98072', '98188', '98014', '98055', '98039'],
      dtype=object)

In [33]:
df.zipcode = df.zipcode.astype(str)

In [36]:
df.zipcode.dtype

dtype('O')

## Step 2: Exploratory Data Analysis of King County Housing Data

## Step 3: Modeling of King County Housing Data with Multivariate Linear Regression

In [13]:
def inspect_df(df=df):
    # display(df.describe().T)
    display(pd.concat([df.head(1).applymap(type), df.head(2)]).T)
    for i, col in enumerate(df):
        print('-------------------------------------')
        print(f'           {col} @ {i}')
        
        percent_nan = round(df[col].isna().sum() / len(df[col]), 4)
        if percent_nan > 0:
            print(f'Uh oh, {col} has {percent_nan} missing values')
        else:
            print('Good, no values missing.')
        if df[col].nunique() < 100:
            print(f"{col} seems to be discrete with {df[col].nunique()} unique values.")
            try:
                plt.figure(figsize=(8,8))
                sns.catplot(x=col, y=df['price'], data=df)
                plt.show()
            except Exception as e: 
                print(e)
                plt.show()
        else:
            print(f"{col} seems to be continuous.")
            plt.figure(figsize=(8,8))
            sns.scatterplot(x=col, y=df['price'], data=df, alpha=.3)
            plt.show()
#         except Exception as e: 
#             print(e)


In [14]:
# give every a row a meaninful number for 'yr_renovated': replace 0 or NaN with 'yr_built' instead
df.loc[df['yr_renovated'].isna(), 'yr_renovated'] = df.loc[df['yr_renovated'].isna(), 'yr_built']
df.loc[df['yr_renovated'] == 0, 'yr_renovated'] = df.loc[df['yr_renovated'] == 0, 'yr_built']
df[['yr_built','yr_renovated']].loc[df.yr_built != df.yr_renovated].shape

KeyError: 'yr_renovated'

In [ ]:
inspect_df()

In [ ]:
df.loc[df['bathrooms'] > 6]

In [ ]:
from sklearn.linear_model import LinearRegression

possible_single_var = ['sqft_living', 'sqft_above', 'sqft_living15']
#for col in possible_single_var:
y = df[['price']]
X = df[['sqft_living']]
X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.3)
linreg = LinearRegression()
model = linreg.fit(X_train, y_train)
predictions = linreg.predict(X_test)

plt.scatter(y_test, predictions)
plt.xlabel(f"True Value")
plt.ylabel(f"Single Variable Linear Predictions")
print("Score:", model.score(X_test, y_test))